### ISSUE TITLE
Eureka Client dependency update to Jersey 2.x

### ISSUE THREAD

**Issue by cforce** [Netflix/eureka#600]
We have to add the spring-boot-starter-jersey to put Jersyey 2.14 on the classpath. The mix of both versions lead to strange NoSuchMethodErrors during application startup! The only solution seems to be to use JAX-RX 1.x (Jersey 1) . This is maybe related to https://github.com/Netflix/eureka/issues/376

**Comment by qiangdavidliu** [comment]
@cforce eureka-client depends on jersey 1.x only and we have not tested (or guarantee) compatibility with jersey 2.x. We also do not have any plans on upgrading to jersey 2.x in the short term.

**Comment by foo4u** [comment]
@qiangdavidliu can we shade the dependency on Jersey 1.x? Causing issues since most other frameworks have moved on to JAX-RS 2.0.

**Comment by tbak** [comment]
We implement a new transport that will allow one to plug-in different HTTP client implementations. It is still on branch, but we will merge this to master probably next week. Once it is done, you can exclude jersey 1.x, and provide equivalent jersey 2.x implementation
Check: https://github.com/Netflix/eureka/tree/refactorings/transport/eureka-client/src/main/java/com/netflix/discovery/shared/transport

If we have time, we might provide a new module (eureka-client-jersey2), that would do exactly this.

**Comment by efenderbosch** [comment]
Would that make it possible to use something other than Jersey? It is just required to be a JAX-RS implementation?

**Comment by tbak** [comment]
For eureka-client, yes this could be anything that implements EurekaHttpClient API. But on the server side we depend heavily on JAX-RS.

**Comment by cforce** [comment]
How its solved now? Do i still need to pimp like this?
 <plugins>
            <plugin>
                <groupId>com.spotify</groupId>
                <artifactId>docker-maven-plugin</artifactId>
            </plugin>
            <plugin>
                <groupId>org.springframework.boot</groupId>
                <artifactId>spring-boot-maven-plugin</artifactId>
                
                <configuration>
                    <requiresUnpack>
                        <dependency>
                            <groupId>com.netflix.eureka</groupId>
                            <artifactId>eureka-core</artifactId>
                        </dependency>
                        <dependency>
                            <groupId>com.netflix.eureka</groupId>
                            <artifactId>eureka-client</artifactId>
                        </dependency>
                    </requiresUnpack>
                </configuration>
            </plugin>
        </plugins>

or 

<dependency>
        <groupId>org.springframework.cloud</groupId>
    <artifactId>spring-cloud-starter-eureka</artifactId>
    <exclusions>                
        <exclusion>
            <artifactId>jsr311-api</artifactId>
            <groupId>javax.ws.rs</groupId>
        </exclusion>
    </exclusions>
</dependency>

**Comment by tbak** [comment]
We are working towards enabling multiple transports in Eureka client. There
is already jersey2 implementation, however DiscoveryClient is still tightly
coupled with jersey1.x. We plan to re-write this code next month, after
holiday season.

**Comment by singerdmx** [comment]
Any updates on Jersey2 version? Thanks!

**Comment by qiangdavidliu** [comment]
@singerdmx we are doing some work that will help with this this quarter, but we can't commit to any timelines right now unfortunately. I will update once we have anything more concrete.

**Comment by clajder** [comment]
Could someone give rough estimate on how long it takes to decouple code from jersey v1? 
Thanks!

**Comment by jasollien** [comment]
Hi guys. Uncopling of jersey1 is essential for us in order to use Eureka. Any update on the progress would be much appreciated.

**Comment by spencergibb** [comment]
It's not a priority for them, but I'm sure they would look at a pull request.

**Comment by saneItchyHog** [comment]
Eagerly awaiting this as well

**Comment by qiangdavidliu** [comment]
@jasollien @saneItchyHog we are taking a more drastic direction w.r.t. the client side of eureka (thinner clients with no jersey dependencies at all), and is working more towards this goal at the moment. We believe that this is a more flexible solution at the end of the day as it addresses other dependency problems as well as the jersey issue.

Us providing a jersey 2.x implementation of the current client is technically feasible, however part of the value of eureka (client and server) is that is it battle tested in very large and varied deployment scenarios at Netflix, and even if we provide a jersey 2.x variant, it will not be used internally and will lack similar levels of production readiness.

Having said the above, the aforementioned (earlier in this thread) new transport is now active, and there is a jersey 2.x skeleton in the eureka-client-jersey2 submodule, so we are definitely more than willing to look at any pull requests to build on top of these.

**Comment by nick-pww** [comment]
@qiangdavidliu I think most of the folks following this would agree that removing the Jersey dependency from your side of things would be the 'best' solution for folks who are using this library. Personally, I don't care what you use for your underlying communications as long it's been well tested and vetted (which is what you are going for), AND doesn't get in the way of our own application development (which Jersey dependencies of any kind have the chance of doing).

tl;dr: 👍 

Look forward to the changes!

**Comment by mattnelson** [comment]
@qiangdavidliu 
> we are taking a more drastic direction w.r.t. the client side of eureka (thinner clients with no jersey dependencies at all), and is working more towards this goal at the moment.

I am starting to look into eureka-client-jersey2 and picking up where you guys left off, but I wanted to put a feeler out to see what progress had been made on the jersey-agnostic client and if there was the possibility of a release candidate being available.

**Comment by qiangdavidliu** [comment]
@mattnelson as part of the new work we are doing a bit more than just a different client, but are also taking advantage of this to rework the data model to update it to the 2016 requirements we have as a company, compared to when eureka was first published (better VPC support for example). Part of the work will include introducing a proxy tier to handle the compatibility with the existing eureka server so I suspect it would not fit with your immediate use case. 

If you are keen to follow up on the existing -jersey2 work that would be greatly appreciated, as I don't think the newer work will completely deprecate all use of the existing client(s). I think it will bring a lot of value to this project.

Thanks.

**Comment by spencergibb** [comment]
@mattnelson we have users (large institutions) that would love have a jersey2 client working.

**Comment by mattnelson** [comment]
Got the PR out for this on #821

**Comment by adoura** [comment]
@spencergibb now jersey 2 compatibility released in eureka 1.6.0 but you planning to upgrade to eureka 1.5.6 in 1.3.0.M1. 
is there possibility to upgrade to 1.6.0 instead ?

**Comment by nmadzharov** [comment]
Considering it has been over a year since the issue was closed and nowadays spring-boot applications with jersey 2 clients are widely-used, what is the recommended approach to use Eureka client in web services which also already use jersey 2 clients? 

For instance, I have struggled to find a recommendation or example on using eureka clients with any spring boot app that utilizes spring-boot-starter-jersey or the likes (jersey 2 dependencies) which are included in most spring examples.
